File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        
        for(int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            
            // Push opening brackets onto the stack
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } 
             else if (c == ')' && !stack.isEmpty() && stack.peek() == '(') {
                stack.pop();
            } 
            else if (c == '}' && !stack.isEmpty() && stack.peek() == '{') {
                stack.pop();
            } 
            else if (c == ']' && !stack.isEmpty() && stack.peek() == '[') {
                stack.pop();
            } 
        else{
            return false;
        }
    }
    return stack.isEmpty();
}
}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pairs = {')': '(', ']': '[', '}': '{'}
    
    #every char in the string
        for ch in s:
            if ch in pairs.values(): #opening braces
                stack.append(ch)
            else:                   #closing braces
                if not stack or stack[-1] != pairs[ch]: #check empty stack or TOP != coressponding closing braces
                    return False
                stack.pop()

        return not stack #returns true since stck becomes empty

        

    
